core/shared/src/main/scala/spire/math/interval/Overlap.scala

```diff
@@ -0,0 +1,65 @@
+package spire.math.interval
+
+import spire.algebra.{Eq, Order}
+import spire.math.Interval
+import spire.math.interval.Overlap.{Disjoint, Equal, Subset}
+
+/**
+  * An ADT that represents overlapping result for any two intervals.
+  */
+sealed abstract class Overlap[A: Order] extends Serializable {
```

Comment:
We should avoid trapping/requiring an `Order` here. You'll need to add an `(implicit ord: Order[A])` parameter to `def join`, but otherwise I think everything should just work.